### PR TITLE
Contextualized logging

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
I've worked on this a couple of month ago and finally had the time to submit it.

The PR enhances the logging by adding the `userID` to every log call making it easy to search for an event related to a user.

#### Ticket Link
None

### CODE REVIEW

**Comment by hanzei** [comment]
@larkox @iomodo Take your time with reviewing this PR. There is no hurry.

File: server/plugin/api.go [hunk 1]
```diff
@@ -105,28 +121,30 @@ func (p *Plugin) initializeAPI() {
...
-	apiRouter.HandleFunc("/milestones", p.extractUserMiddleWare(p.getMilestones, ResponseTypePlain)).Methods(http.MethodGet)
-	apiRouter.HandleFunc("/assignees", p.extractUserMiddleWare(p.getAssignees, ResponseTypePlain)).Methods(http.MethodGet)
-	apiRouter.HandleFunc("/repositories", p.extractUserMiddleWare(p.getRepositories, ResponseTypePlain)).Methods(http.MethodGet)
-	apiRouter.HandleFunc("/settings", p.extractUserMiddleWare(p.updateSettings, ResponseTypePlain)).Methods(http.MethodPost)
-	apiRouter.HandleFunc("/user", p.extractUserMiddleWare(p.getGitHubUser, ResponseTypeJSON)).Methods(http.MethodPost)
-	apiRouter.HandleFunc("/issue", p.extractUserMiddleWare(p.getIssueByNumber, ResponseTypePlain)).Methods(http.MethodGet)
-	apiRouter.HandleFunc("/pr", p.extractUserMiddleWare(p.getPrByNumber, ResponseTypePlain)).Methods(http.MethodGet)
+	oauthRouter.HandleFunc("/connect", p.checkAuth(p.attachContext(p.connectUserToGitHub), ResponseTypePlain)).Methods(http.MethodGet)
+	oauthRouter.HandleFunc("/complete", p.checkAuth(p.attachContext(p.completeConnectUserToGitHub), ResponseTypePlain)).Methods(http.MethodGet)
```
Comment (iomodo) [inline server/plugin/api.go:125]: I don't like that we have to do `r.Header.Get("Mattermost-User-ID")` twice. Could we transfer userID between middlewares?

**Review by iomodo — COMMENTED** [review]
Wonderful! Just couple of questions.

File: server/plugin/api.go [hunk 2]
```diff
@@ -162,7 +180,57 @@ func (p *Plugin) extractUserMiddleWare(handler HTTPHandlerFuncWithUser, response
...
+}
+
+func (p *Plugin) createContext(_ http.ResponseWriter, r *http.Request) *Context {
+	userID := r.Header.Get("Mattermost-User-ID")
+
+	logger := logger.New(p.API).With(logger.LogContext{
+		"userid": userID,
+	})
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
[the remark this answers is not in the document]
Comment (hanzei) [reply server/plugin/api.go:194]: Best practice. HTTP requests should time out after some seconds. 0/5 if 5 is the right number here
.

File: server/plugin/api.go (hunk 1 above)
Comment (hanzei) [reply server/plugin/api.go:125]: A HTTP header is static information to read from. My using `r.Header.Get("Mattermost-User-ID")` twice the code is sure to use the single source of truth every time.

**Review by iomodo — APPROVED** [review]

**Review by larkox — APPROVED** [review]

**Comment by hanzei** [comment]
@DHaussermann I'm not sure if there is much to test here. The logging is only triggered when a call to the github api fails. You can enforce that by e.g. disconnecting for the internet.

**Comment by hanzei** [comment]
@DHaussermann Would you be fine testing this as part of the release process?

**Review by DHaussermann — APPROVED** [review]
Thanks @hanzei 

Yes, I'm okay with testing this logging enhancement post-merge as part of release testing. 

Please merge.

**Pull request merged**